### ISSUE TITLE
Possible Deadlock in Broker-based Zeek Clusters

### ISSUE THREAD

**Issue by Neverlord** [zeek/broker#165]
In the original design for Broker, we only allowed loop-free topologies. Based on that assumption, we settled on stream-based communication backend to get overload protection via back-pressure. The backend still runs on the assumption of loop-free topologies, but Zeek does not use Broker in this way.

Later additions to Broker introduced the forwarding flag to basically cheat Broker to have loops in the topology while *logically* cutting off all message paths that would go through loops. Zeek does exactly that: it spins up manager, proxy, worker and logger and then forms a full mesh between them while turning forwarding off everywhere. The idea is that the loops exist, but that no message in practice travels more than one hop.

But the loops still exist. The stream-based backend does not know about the forwarding flag and still relies on the assumption that topologies are loop-free.

With recent `master`, the `broker-cluster-benchmark` tool occasionally runs into a deadlock (@timwoj thanks again for providing the recording that reproduced this!). We've migrated to CAF 0.18 recently, so I was worried that it might be CAF bug. After debugging this for a while, I've finally managed to get a trace log and reproduce what's happening. With the benchmarking tool, all Broker endpoints start sending as fast as they can right away. The deadlock manifests itself when:

- the manager waits for credit from the proxy
- the proxy waits for credit from the worker
- the worker waits for credit from the manager

We know that the proxy should give new credit to the manager, because no forwarding to the worker will ever take place. However, the streams don't know that. They have no credit downstream and so they won't grant more credit upstream.

We didn't notice this deadlock before, because CAF 0.17 uses buffers the size of the moon. The oversized buffers made this condition less likely to occur, but once we hit 0 credit on the "wrong" paths, it's game over for Broker. Increasing the buffer sizes again may work as a mitigation strategy, but actually fixing the deadlock won't be straightforward.

I think we still want to keep the back-pressure, but the topic-based messaging logic doesn't fit naturally onto streams any more after allowing loops in the topology. We still want to have streams from local publishers / to local subscribers in order to be able to slow down fast publishers or signal overloads to peers if local subscribers become overloaded. With our loop-free topologies assumption, we simply had one big stream with many paths (to local publishers/subscribers as well as to peers). If we break this up into multiple streams, then we must make sure to include feedback between the streams in order to preserve the back-pressure while making sure to include only the necessary feedback to avoid "false dependencies" that lead to the current deadlock.

**Comment by Neverlord** [comment]
So here's what I propose to fix the issue in a way that's also going to work with the ALM version of Broker. Rather than having one big stream with `n` paths, we have `n` streams all connected through a central (virtual) queue:

![redesign](https://user-images.githubusercontent.com/651128/103284048-759f3b80-49da-11eb-9bef-e5a58c072e5f.png)

On its own, that wouldn't resolve the issue if we would only use the size of the queue to calculate the credit we can communicate upstream. So there's one more step: keeping track of where data came from. For that, I propose these interfaces (pseudo code):

```c++
struct source {
  void reclaim(what: item*);
  std::vector<item*> free_list;
};

struct item {
  source* origin;
  size_t ref_count;
  broker::message data;
};
```

The gist of this idea is to allow each source to inject `n` messages into the system, where `n` is some upper bound defined by the credit policy. Initially, a source has `n` items in its `free_list`. Whenever we receive data from a source, we wrap it into an `item` and push it into the (virtual) central queue with a `ref_count` of 1. Whenever the item gets enqueued into another queue, we increment `ref_count`. Once the item got consumed, we decrement the `ref_count`. Once it hits 0, we call `reclaim` on the source of the item to add it back to the `free_list`.

The central queue only exists for providing a central dispatching logic and it does not store any item on its own. Whenever a source pushes to it, the queue pushes the item to all connected queues with matching subscriptions (plus checking that the destination is not the same as the `origin`).

With this in place, we can simply communicate the size of the `free_list` upstream as credit. This solves our deadlock by keeping track of the actual dependencies in the system. Here's an example from the Zeek cluster:

- The `proxy` receives messages from the `manager` for some topic `t0` with local subscribers.
- Local publishers at the `proxy` push messages to another topic `t1` that we will send to the `worker`.
- No message for `t0` ever ends up in the queue for the `worker`. With the new design, this means no `item` with origin `manager` touches a queue connected to the `worker` and thus the `worker` can no longer affect credit decisions for the `manager`. That is unless there are *actual* loops in the topology, which we don't support. The ALM branch safeguards against this by deploying source routing to make sure messages travel on deterministic paths even if the topology contains loops.

@rsmmr, @jsiwek are you on board with this design change or do you see flaws in it?

**Comment by Neverlord** [comment]
I've started working on this design in `issue/165` to see if there are any road blockers. The tricky bit is to get this hooked up into CAF's credit management for streams, so I've started with that. I went with a custom `stream_manager` design that takes  few shortcuts to keep things simple (e.g. only allowing one in- or outbound path). Looks promising so far, most of the remaining work is integrating the new scaffold into the core actor.

**Comment by jsiwek** [comment]
> We didn't notice this deadlock before, because CAF 0.17 uses buffers the size of the moon. The oversized buffers made this condition less likely to occur, but once we hit 0 credit on the "wrong" paths, it's game over for Broker. Increasing the buffer sizes again may work as a mitigation strategy, but actually fixing the deadlock won't be straightforward.
> 
> I think we still want to keep the back-pressure

Be patient with this, but wondering if we can do the exercise of revisiting/understanding what problems are addressed by the backpressure mechanism.

First, was curious if you had more description of what the CAF 0.17 behavior was: what does "buffers the size of the moon" mean ?  Did we likely even have any backpressure occurring ?  If no, that might be a flag that it's not even that critical of a feature at the moment (given no apparent reports of problems due to lack of it working properly).

Generally, what does the backpressure accomplish ?  I think it hypothetically results in a more efficient distribution of memory under high loads such that any given node is less likely to get overwhelmed in a short-term/transient/bursty sense, but for long-term/steady-state where senders (instead of receivers) are still going to have to buffer excessive messages (since there's currently no policy/logic to drop anything), does backpressure by itself prevent nodes from eventually falling over ?  Thinking "no", but also that may be fine if it's a matter of "it's still desirable that backpressure lets them last longer before falling over" and/or it becomes more important at larger scales and we have done some simulation/testing to show this actually helping in the intended way at whatever scale.

> @rsmmr, @jsiwek are you on board with this design change or do you see flaws in it?

Nothing else contrary from me, but also my understanding of the details of how the streaming backpressure/crediting logic works may be too limited (or even wrong) to give good feedback.  Can a more detailed illustration of it be made ?  Besides helping this discussion/review, could add it directly to Broker's docs (a "for developers" section) for reference by future people.  E.g. helps to have aid/visuals in answering these types of questions:

* What's the meaning/relationship between these terms: stream, path, peer, subscriber/publisher, topic
* When are credit decisions/accounting done and also where are they done (e.g. in context/relationship to above terms)


Lastly, just an impression of the bug's severity/timeline: seems important that a fix for this get into the Zeek 4.0 release (let me know if there's other thoughts/concerns).

**Comment by rsmmr** [comment]
After talking to @Neverlord I see two options for 4.0: 

- Integrate the changes being worked on now to fix the problem
- Work around the problem with some version of disabling the backpressure/credit system (easiest might be configuring a huge buffer size).

I think the latter would feel most comfortable to me given the timeline, but more opinions appreciated.

**Comment by Neverlord** [comment]
>     * Work around the problem with some version of disabling the backpressure/credit system (easiest might be configuring a huge buffer size).

That workaround would boil down to this patch:

```diff
diff --git a/src/configuration.cc b/src/configuration.cc
index e29ef26..7112053 100644
--- a/src/configuration.cc
+++ b/src/configuration.cc
@@ -99,6 +99,8 @@ configuration::configuration(skip_init_t) {
   // Broker didn't load the MM module yet. Use `put` to suppress the 'failed to
   // set config parameter' warning on the command line.
   put(content, "caf.middleman.app-identifiers", std::move(ids));
+  // Use freakingly large buffers (up to  512MB) as workaround for #165.
+  put(content, "caf.stream.size-based-policy.buffer-capacity", 536870912);
   // Turn off all CAF output by default.
   std::vector<std::string> excluded_components{"caf", "caf_io", "caf_net",
                                                "caf_flow", "caf_stream"};
```

I couldn't reproduce the deadlock with this buffer size. However, this *does* come at the expense of performance due to buffer bloat. At least for the benchmark.

> First, was curious if you had more description of what the CAF 0.17 behavior was: what does "buffers the size of the moon" mean ? Did we likely even have any backpressure occurring ? If no, that might be a flag that it's not even that critical of a feature at the moment (given no apparent reports of problems due to lack of it working properly).

CAF ≤ 0.17 had a somewhat convoluted way of computing the maximum amount of credit. The idea was to measure how long it takes to process a batch and then estimate how much items the actor would need to run at a constant pace. However, measuring this time and extrapolating based on that was way to imprecise. In practice, CAF would hugely overestimate how many items an actor should request from upstream for any given time interval. The old credit policy also didn't result in particularly stable system behavior because the estimates fluctuated too heavily. In numbers: I measured Broker peaking somewhere around 250MB when running the `broker-benchmark`. Here's a Grafana dashboard visualizing 15 Minutes worth of CAF metrics for a `broker-benchmark` run I did back in October.

![image](https://user-images.githubusercontent.com/651128/103816126-c5c66f80-5064-11eb-8165-9a1a291c2f5f.png)

With the new credit policy for CAF 0.18, I later measured the otherwise unchanged setup with ~130k messages / second (up from ~20k/s) while staying below 15MB resident set size. The new policy is size-based and much simpler. It allows users to configure a desired upper bound (default is 64kB) and then samples the size of incoming data (by computing how much bytes the data would occupy when serialized).

> Generally, what does the backpressure accomplish ?

Most importantly, it gives feedback to publishers. Right now, Zeek doesn't tap into this. However, Broker could expose the current load to Zeek to allow it to deal with congestion / overloads. Zeek could, for example, drop some less important log events or delay them to allow more important communication to go through.

For Broker itself, the streaming backend achieves a couple of things. Slowing down fast senders, but it also ensures fairness. When receiving data from multiple sources (local publishers and peers), the streams make sure that each source gets its turn (the algorithm CAF uses internally is called [deficit round robin](https://en.wikipedia.org/wiki/Deficit_round_robin)). If Broker would use regular actor communication here, then it would have to process messages in the order they arrived in. The core actor would likely become very unresponsive once it has some serious load from publishers.

> does backpressure by itself prevent nodes from eventually falling over ? 

There's a bit more to streams than "just" the backpressure. In comination with the other aspects (multiplexing, baching, etc), though: yes. Instead of blowing up subscribers, Broker slows down the fast publisher. Since Zeek currently does not react to overloads in a meaningful/intelligent way, this at least makes sure that only one Broker endpoint falls over: the one that tries to publish more than the subscribers can swallow.

> Can a more detailed illustration of it be made ?

I'll make sure to provide more details when I file the PR. That will end up in the developer guide as well. This guide lives in the ALM branch, though, so I'll update it when integrating the new code to the ALM backend.

> Lastly, just an impression of the bug's severity/timeline: seems important that a fix for this get into the Zeek 4.0 release (let me know if there's other thoughts/concerns).

I'm working on getting the patch ready for PR. For a bugfix, this is touching quite a bit of code, though. So whether you pull in the PR or apply the above workaround patch for next 4.0 RC / release is up to what you feel most comfortable with. 🙂
